<a href="https://colab.research.google.com/github/VinayaGajula/CODSOFT/blob/main/CODSOFT_TASK_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**TASK 5: CREDIT CARD DETECTION**


*   DEVELOPED BY: G. Vinaya Lakshmi
   



# Importing Required Libraries

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Loading DataSet to a Pandas DataFrame

In [ ]:
df= pd.read_csv('/creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [30]:
# reading last 5 rows from the dataframe
df.tail()

Time        V1        V2        V3        V4        V5        V6  \
120896  75998  1.501845 -1.193147  0.066789 -1.600847 -1.259865 -0.268558   
120897  75999 -0.340455  1.201756  1.314094  0.053134  0.111616 -0.959891   
120898  75999 -1.089161  0.115364  1.501978  0.700866  0.493027 -0.220268   
120899  76000  0.213729  0.623136 -0.259558 -0.152127  2.262088  3.579712   
120900  76001 -0.367988  1.108999  0.787889  1.043825 -0.336914 -0.554318   

              V7        V8        V9  ...       V21       V22       V23  \
120896 -1.122205  0.040104 -1.795060  ... -0.226430 -0.409194 -0.074263   
120897  0.780107 -0.128473 -0.541652  ... -0.252458 -0.609959 -0.025552   
120898  0.744468  0.128396 -0.983044  ...  0.391619  0.776285  0.179390   
120899 -0.187220  0.877580 -0.697998  ... -0.193588 -0.663299  0.088675   
120900  0.307948  0.402319 -0.651386  ...  0.226379  0.551456  0.058382   

             V24       V25       V26       V27       V28  Amount  Class  
120896 -0.565449  0.404350 -0.168445  0.010981  0.001626   30.00    0.0  
120897  0.360654 -0.127871  0.068601  0.251848  0.100380    3.67    0.0  
120898  0.218712 -0.009381 -0.353213  0.082478  0.143270  130.80    0.0  
120899  1.005326 -0.676412  0.251661  0.105658  0.088634    1.29    0.0  
120900  0.394346       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [ ]:
#Infromation about dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120901 entries, 0 to 120900
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    120901 non-null  int64  
 1   V1      120901 non-null  float64
 2   V2      120901 non-null  float64
 3   V3      120901 non-null  float64
 4   V4      120901 non-null  float64
 5   V5      120901 non-null  float64
 6   V6      120901 non-null  float64
 7   V7      120901 non-null  float64
 8   V8      120901 non-null  float64
 9   V9      120901 non-null  float64
 10  V10     120901 non-null  float64
 11  V11     120901 non-null  float64
 12  V12     120901 non-null  float64
 13  V13     120901 non-null  float64
 14  V14     120901 non-null  float64
 15  V15     120901 non-null  float64
 16  V16     120901 non-null  float64
 17  V17     120901 non-null  float64
 18  V18     120901 non-null  float64
 19  V19     120901 non-null  float64
 20  V20     120901 non-null  float64
 21  V21     12

Checking Missing values

In [ ]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

we can observe that there are no missing values and hence we can continue.

In [ ]:
df['Class'].value_counts()

0.0    120651
1.0       249
Name: Class, dtype: int64

distribute transactions on the basis of class
for Class=0 we say those are normal transactions i.e., legit transactions (120651)
whereas Class=1 are fradulent transactions(249)

This dataset is highly Unbalanced.

In [ ]:
#seperating data for analysis.
legit= df[df.Class== 0]
fraud=df[df.Class==  1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(120651, 31)
(249, 31)


statistical measures of data

In [ ]:
legit.Amount.describe()

count    120651.000000
mean         94.185676
std         255.770139
min           0.000000
25%           6.885000
50%          24.990000
75%          84.815000
max       19656.530000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()

count     249.000000
mean      119.180482
std       251.160611
min         0.000000
25%         1.000000
50%        11.390000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

comparing the values for both transactions (based on mean value)

In [ ]:
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    47510.269886 -0.242842 -0.016464  0.695109  0.143967 -0.275556   
1.0    40280.433735 -5.846600  4.076891 -7.453278  4.641314 -4.167167   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.091375 -0.101951  0.057796 -0.073721  ...  0.042749 -0.038022   
1.0   -1.511639 -6.201243  1.559279 -2.696182  ...  0.259111  1.324185   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.113385 -0.035057  0.011751  0.130599  0.027544 -0.000211  0.001646   
1.0   -0.314840 -0.103567 -0.108411  0.199329  0.074713  0.522683  0.076517   

           Amount  
Class              
0.0     94.185676  
1.0    119.180482  

[2 rows x 30 columns]

Under-Sampling

Build a sample dataset having similar distribution of Legit and Fraudulent transactions

number of Fraudulent Transactions=249

therefore add random 249 transactions from normal transactions and add it to Fraudulent transactions so as to make uniform distribution of data in both the cases

In [ ]:
legit_sample= legit.sample(n=249)

Concatinating Two DataFrames.

In [ ]:
new_df= pd.concat([legit_sample,fraud],axis=0)

In [ ]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
44195  41840 -0.612955  1.066750  1.670531  0.059486 -0.658606 -0.800321   
85126  60618  1.114325 -1.019216 -0.326087 -0.431998 -0.840991 -0.703095   
9264   13340 -0.690693  0.975959  1.317698 -0.012341  0.854941  1.098072   
49673  44158 -1.242261  1.174146  2.236598  2.859563 -0.282746  0.766849   
16521  27891  1.079749 -0.503070  1.231823  0.130393 -1.309191 -0.246212   

             V7        V8        V9  ...       V21       V22       V23  \
44195  1.431650 -0.611567 -0.293230  ...  0.001830  0.134717 -0.123331   
85126 -0.032779 -0.285653 -0.728653  ... -0.385516 -0.904172 -0.216926   
9264   0.420140  0.306975  1.420890  ... -0.395011 -0.491027  0.119205   
49673 -0.151577  0.260983 -0.196646  ... -0.196109 -0.065540 -0.246472   
16521 -0.729543  0.160952  0.803688  ... -0.027133 -0.012243  0.035714   

            V24       V25       V26       V27       V28  Amount  Class  
44195  0.953568 -0.200622  0.147369 -0.470860 -0.208672  111.18    0.0  
85126 -0.094177  0.431027  1.107851 -0.103981  0.021508  177.13    0.0  
9264  -1.115370 -0.519210  0.143270  0.233460 -0.092122   18.99    0.0  
49673 -0.117812  0.319460  0.291651  0.396989  0.307908   41.49    0.0  
16521  0.615447  0.041539  0.955731 -0.044963  0.015564   50.00    0.0  

[5 rows x 31 columns]

In [ ]:
new_df['Class'].value_counts()


0.0    249
1.0    249
Name: Class, dtype: int64

uniformly distributed data is formed with 249 normal and 249 Fraudulent transactions.

In [ ]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    47361.096386 -0.408509 -0.102529  0.840053  0.111908 -0.173793   
1.0    40280.433735 -5.846600  4.076891 -7.453278  4.641314 -4.167167   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.104122 -0.096663  0.081965 -0.149686  ...  0.037657  0.013157   
1.0   -1.511639 -6.201243  1.559279 -2.696182  ...  0.259111  1.324185   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.177892  0.041944  0.121535  0.066442  0.004409  0.016296 -0.020057   
1.0   -0.314840 -0.103567 -0.108411  0.199329  0.074713  0.522683  0.076517   

           Amount  
Class              
0.0     90.711245  
1.0    119.180482  

[2 rows x 30 columns]

We can observe that there is not much difference in mean values before and after adding transaction values into fraudulent transactions.(i.e., nature of dataset is not changed which helps our machine for easy detection of fraud transactions.)

Splitting the data into Features and Targets(Class)

In [ ]:
X = new_df.drop(columns='Class', axis=1)
Y = new_df['Class']

In [ ]:
print(X)

         Time        V1        V2        V3        V4        V5        V6  \
44195   41840 -0.612955  1.066750  1.670531  0.059486 -0.658606 -0.800321   
85126   60618  1.114325 -1.019216 -0.326087 -0.431998 -0.840991 -0.703095   
9264    13340 -0.690693  0.975959  1.317698 -0.012341  0.854941  1.098072   
49673   44158 -1.242261  1.174146  2.236598  2.859563 -0.282746  0.766849   
16521   27891  1.079749 -0.503070  1.231823  0.130393 -1.309191 -0.246212   
...       ...       ...       ...       ...       ...       ...       ...   
118308  75033 -0.430330  0.985633  0.645789  0.317131  0.616332 -1.347462   
119714  75556 -0.734303  0.435519 -0.530866 -0.471120  0.643214  0.713832   
119781  75581 -2.866364  2.346949 -4.053307  3.983359 -3.463186 -1.280953   
120505  75851 -4.793667  3.418911 -5.074445  4.035987 -3.527875 -1.923242   
120837  75978 -5.140723  3.568751 -5.896245  4.164720 -4.091193 -1.989960   

              V7        V8        V9  ...       V20       V21       V22  \


In [ ]:
print(Y)

44195     0.0
85126     0.0
9264      0.0
49673     0.0
16521     0.0
         ... 
118308    1.0
119714    1.0
119781    1.0
120505    1.0
120837    1.0
Name: Class, Length: 498, dtype: float64


**Applying Train_Test Split method**

split the data into testing data(20%) and training data(80%)
 and stratifying data according to 'Y'

In [ ]:
X_train, X_test, Y_train, Y_test= train_test_split(X,Y, test_size=0.2, stratify=Y,random_state=2)

In [ ]:
#checking sizes of dataset
print(X.shape, X_train.shape, X_test.shape)

(498, 30) (398, 30) (100, 30)


Model Training

**Logistic Regression**

In [ ]:
model= LogisticRegression()

training logistic regression model with training data

In [ ]:
model.fit(X_train, Y_train)

LogisticRegression()

Model Evaluation

Finding Accuracy (accuracy score function)

In [ ]:
X_train_prediction = model.predict(X_train)
training_data_accuracy =  accuracy_score(X_train_prediction, Y_train)

In [31]:
print('Accuracy on training data', training_data_accuracy)

Accuracy on training data 0.957286432160804


In [ ]:
X_test_prediction =  model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [ ]:
print('Accuracy score on test data', test_data_accuracy)

Accuracy score on test data 0.95


**CONCLUSION**

 As the DataSet is distributed uniformly now, and

 accuracy score of both training data and testing data is almost similar i.e., data set is not overfitted and underfitted

 therefore, credit card fraud detection will be done more accurately by machine.